### PR TITLE
Refactor HIR item-like traversal (part 1)

### PR DESCRIPTION
**Pull request by kckeiks**
Issue  #95004

- Create hir_crate_items query which traverses tcx.hir_crate(()).owners to return a hir::ModuleItems
- use tcx.hir_crate_items in tcx.hir().items() to return an iterator of hir::ItemId
- use tcx.hir_crate_items to introduce a tcx.hir().par_items(impl Fn(hir::ItemId)) to traverse all items in parallel;

Signed-off-by: Miguel Guarniz <mi9uel9@gmail.com>

cc @cjgillot

### CODE REVIEW

File: compiler/rustc_middle/src/query/mod.rs [hunk 1]
```diff
@@ -1966,4 +1966,11 @@ rustc_queries! {
         eval_always
         desc { "computing the backend features for CLI flags" }
     }
+
+    /// All items in the crate.
+    query hir_crate_items(_: ()) -> rustc_middle::hir::ModuleItems {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/query/mod.rs:1971]: Could you place it with the other HIR-related queries?  For instance above hir_module_items?

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 2]
```diff
@@ -1310,3 +1306,64 @@ pub(super) fn hir_module_items(tcx: TyCtxt<'_>, module_id: LocalDefId) -> Module
...
+    };
+
+    tcx.hir().walk_toplevel_module(&mut collector);
+
+    let CrateCollector { submodules, items, trait_items, impl_items, foreign_items, .. } =
+        collector;
+
+    return ModuleItems {
+        submodules: submodules.into_boxed_slice(),
+        items: items.into_boxed_slice(),
```
Comment (cjgillot) [inline compiler/rustc_middle/src/hir/map/mod.rs:1327]: Reading the code, I'm not sure that `items` starts with `CRATE_DEF_ID`.  It should.

**Review by cjgillot — COMMENTED** [review]
Thanks @kckeiks.  This is a good start.  I have a few remarks/questions.

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 3]
```diff
@@ -1310,3 +1306,64 @@ pub(super) fn hir_module_items(tcx: TyCtxt<'_>, module_id: LocalDefId) -> Module
...
+    impl<'hir> Visitor<'hir> for CrateCollector<'hir> {
+        type NestedFilter = nested_filter::All;
+
+        fn nested_visit_map(&mut self) -> Self::Map {
+            self.tcx.hir()
+        }
+
+        fn visit_item(&mut self, item: &'hir Item<'hir>) {
+            self.items.push(item.item_id());
+            intravisit::walk_item(self, item)
```
[the remark this answers is not in the document]
Comment (kckeiks) [reply compiler/rustc_middle/src/hir/map/mod.rs:1351]: Yes, I can update this method. I think `for_each_module` it's already there.
https://github.com/rust-lang/rust/blob/6a9080b25e73d26aae94c3f6a13b31de58e66b5a/compiler/rustc_middle/src/hir/map/mod.rs#L679

File: compiler/rustc_middle/src/hir/map/mod.rs (hunk 2 above)
Comment (kckeiks) [reply compiler/rustc_middle/src/hir/map/mod.rs:1327]: Now that you mention it. Looking at where the walk begins in `walk_toplevel_module`, it seems  like `CRATE_DEF_ID` does not get added to `items`. Am I reading that right?

File: compiler/rustc_middle/src/hir/map/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kckeiks) [reply compiler/rustc_middle/src/hir/map/mod.rs:1355]: I see what you mean. I will update `for_each_module` now that I'm collecting modules.

**Comment by kckeiks** [comment]
> The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5838314830?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/f213ed6ea86cc843788568b7976ac5a6355bbca1/checks/5838314830/logs)
> Click to see the possible cause of the failure (guessed by this bot)

Updated clippy test.

**Comment by kckeiks** [comment]
@cjgillot I imagine that the performance regression comes from the traversal that the visitor does in `hir_crate_items`.

File: compiler/rustc_hir/src/target.rs [hunk 4]
```diff
@@ -130,6 +131,29 @@ impl Target {
         }
     }
 
+    pub fn from_def_kind(def_kind: DefKind) -> Target {
```
Comment (cjgillot) [inline compiler/rustc_hir/src/target.rs:134]: This is a great idea! I wonder if we should somehow merge `DefKind` and `Target` in a follow-up PR.

File: compiler/rustc_hir/src/target.rs [hunk 5]
```diff
@@ -130,6 +131,29 @@ impl Target {
...
+            DefKind::GlobalAsm => Target::GlobalAsm,
+            DefKind::TyAlias => Target::TyAlias,
+            DefKind::OpaqueTy => Target::OpaqueTy,
+            DefKind::Enum => Target::Enum,
+            DefKind::Struct => Target::Struct,
+            DefKind::Union => Target::Union,
+            DefKind::Trait => Target::Trait,
+            DefKind::TraitAlias => Target::TraitAlias,
+            DefKind::Impl => Target::Impl,
+            _ => panic!("impossible case reached"),
```
Comment (cjgillot) [inline compiler/rustc_hir/src/target.rs:153]: Why is it impossible?

File: compiler/rustc_incremental/src/persist/dirty_clean.rs [hunk 6]
```diff
@@ -147,7 +146,28 @@ pub fn check_dirty_clean_annotations(tcx: TyCtxt<'_>) {
 
     tcx.dep_graph.with_ignore(|| {
         let mut dirty_clean_visitor = DirtyCleanVisitor { tcx, checked_attrs: Default::default() };
-        tcx.hir().visit_all_item_likes(&mut dirty_clean_visitor);
+
+        let crate_items = tcx.hir_crate_items(());
+
+        for id in crate_items.items() {
+            let item = tcx.hir().item(id);
+            dirty_clean_visitor.check_item(item.def_id, item.span);
```
Comment (cjgillot) [inline compiler/rustc_incremental/src/persist/dirty_clean.rs:154]: We don't need to query the HIR to get the span. We can get it using
```rust
let span = tcx.def_span(id.def_id);
dirty_clean_visitor.check_item(id.def_id, span);
```
(or, even better, only pass the `LocalDefId` and have `check_item` call `def_span`)

File: compiler/rustc_interface/src/proc_macro_decls.rs [hunk 7]
```diff
@@ -1,13 +1,18 @@
...
-    tcx.hir().visit_all_item_likes(&mut finder);
+
+    for id in tcx.hir().items() {
+        let attrs = finder.tcx.hir().attrs(id.hir_id());
+        if finder.tcx.sess.contains_name(attrs, sym::rustc_proc_macro_decls) {
+            finder.decls = Some(id.hir_id());
+        }
+    }
 
     finder.decls.map(|id| tcx.hir().local_def_id(id))
```
Comment (cjgillot) [inline compiler/rustc_interface/src/proc_macro_decls.rs:17]: `tcx.hir().local_def_id(it.hir_id())` is equal to `id.def_id`. You can store an `Option<LocalDefId>` directly to simplify.

File: compiler/rustc_interface/src/proc_macro_decls.rs [hunk 8]
```diff
@@ -1,13 +1,18 @@
...
 use rustc_span::symbol::sym;
 
 fn proc_macro_decls_static(tcx: TyCtxt<'_>, (): ()) -> Option<LocalDefId> {
     let mut finder = Finder { tcx, decls: None };
-    tcx.hir().visit_all_item_likes(&mut finder);
+
+    for id in tcx.hir().items() {
+        let attrs = finder.tcx.hir().attrs(id.hir_id());
+        if finder.tcx.sess.contains_name(attrs, sym::rustc_proc_macro_decls) {
+            finder.decls = Some(id.hir_id());
```
Comment (cjgillot) [inline compiler/rustc_interface/src/proc_macro_decls.rs:13]: Great, we don't even need to look at the HIR! (Attributes are a separate query.)

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 9]
```diff
@@ -1783,10 +1783,31 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
...
         let tcx = self.tcx;
-        let mut visitor = ImplsVisitor { tcx, impls: FxHashMap::default() };
-        tcx.hir().visit_all_item_likes(&mut visitor);
+        let mut fx_hash_map: FxHashMap<DefId, Vec<(DefIndex, Option<SimplifiedType>)>> =
+            FxHashMap::default();
+
+        for id in tcx.hir().items() {
+            match tcx.hir().def_kind(id.def_id) {
+                DefKind::Impl => {
+                    let item = tcx.hir().item(id);
```
Comment (cjgillot) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1792]: I'm not sure you need `item` here. Can you directly use `id.def_id` below?

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 10]
```diff
@@ -159,12 +158,12 @@ impl<'hir> Map<'hir> {
...
     }
 
-    pub fn items(self) -> impl Iterator<Item = &'hir Item<'hir>> + 'hir {
-        let krate = self.krate();
-        krate.owners.iter().filter_map(|owner| match owner.as_owner()?.node() {
-            OwnerNode::Item(item) => Some(item),
-            _ => None,
-        })
+    pub fn items(self) -> impl Iterator<Item = ItemId> + 'hir {
+        self.tcx.hir_crate_items(()).items.iter().map(|id| *id)
```
Comment (cjgillot) [inline compiler/rustc_middle/src/hir/map/mod.rs:162]:
```suggestion
        self.tcx.hir_crate_items(()).items.iter().copied()
```

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 11]
```diff
@@ -159,12 +158,12 @@ impl<'hir> Map<'hir> {
...
+    }
+
+    pub fn par_items(self, f: impl Fn(ItemId) + Sync + Send) {
+        par_for_each_in(self.tcx.hir_crate_items(()).items.to_vec(), f);
```
Comment (cjgillot) [inline compiler/rustc_middle/src/hir/map/mod.rs:166]: `to_vec` allocates a new vector needlessly.

```suggestion
        par_for_each_in(&self.tcx.hir_crate_items(()).items[..], |id| f(*id));
```

Comment (cjgillot) [reply compiler/rustc_middle/src/hir/map/mod.rs:166]: Could you do something similar to `par_for_each_module` and only keep one version?

File: compiler/rustc_middle/src/hir/mod.rs [hunk 12]
```diff
@@ -45,6 +45,21 @@ pub struct ModuleItems {
     foreign_items: Box<[ForeignItemId]>,
 }
 
+impl ModuleItems {
+    pub fn items(&self) -> impl Iterator<Item = ItemId> {
+        self.items.to_vec().into_iter()
```
Comment (cjgillot) [inline compiler/rustc_middle/src/hir/mod.rs:50]: `to_vec` allocates needlessly. Likewise in the other methods.
```suggestion
        self.items.iter().copied()
```

File: compiler/rustc_middle/src/ty/print/pretty.rs [hunk 13]
```diff
@@ -2682,8 +2681,13 @@ define_print_and_forward_display! {
...
     let hir = tcx.hir();
-    for item in hir.items() {
-        if item.ident.name.as_str().is_empty() || matches!(item.kind, ItemKind::Use(_, _)) {
+    for id in hir.items() {
+        if matches!(hir.def_kind(id.def_id), DefKind::Use) {
+            continue;
+        }
+
+        let item = hir.item(id);
+        if item.ident.name.as_str().is_empty() {
```
Comment (cjgillot) [inline compiler/rustc_middle/src/ty/print/pretty.rs:2690]: We can avoid accessing the interner by comparing the `Symbol` directly.
```suggestion
        if item.ident.name == kw::Empty {
```

File: compiler/rustc_monomorphize/src/collector.rs [hunk 14]
```diff
@@ -1139,87 +1147,74 @@ struct RootCollector<'a, 'tcx> {
...
+                                );
+
+                                let ty =
+                                    Instance::new(item.def_id.to_def_id(), InternalSubsts::empty())
+                                        .ty(self.tcx, ty::ParamEnv::reveal_all());
+                                visit_drop_use(self.tcx, ty, true, DUMMY_SP, self.output);
+                            }
+                        }
                     }
+                    _ => {}
```
Comment (cjgillot) [inline compiler/rustc_monomorphize/src/collector.rs:1173]:
```suggestion
                    _ => bug!(),
```

File: compiler/rustc_monomorphize/src/collector.rs [hunk 15]
```diff
@@ -1139,87 +1147,74 @@ struct RootCollector<'a, 'tcx> {
...
                 // but even just declaring them must collect the items they refer to
-                if let Ok(val) = self.tcx.const_eval_poly(item.def_id.to_def_id()) {
+                if let Ok(val) = self.tcx.const_eval_poly(id.def_id.to_def_id()) {
                     collect_const_value(self.tcx, val, &mut self.output);
                 }
             }
-            hir::ItemKind::Fn(..) => {
-                self.push_if_root(item.def_id);
+            DefKind::Impl => {
+                let item = self.tcx.hir().item(id);
```
Comment (cjgillot) [inline compiler/rustc_monomorphize/src/collector.rs:1200]: Nit: we can call `tcx.hir().item` inside the `if` branch below.

**Review by cjgillot — COMMENTED** [review]
This is taking a good shape, thanks @kckeiks. I left a few (many?) nits. Let's see what perf says about it.

File: compiler/rustc_hir/src/target.rs (hunk 5 above)
Comment (kckeiks) [reply compiler/rustc_hir/src/target.rs:153]: I'm sorry, this is an oversight. If this is used with `DefKind`s derived from `ItemId`s then I think it is not possible to see the other variants of `DefKind` here but I am not sure about `DefKind`s from `ImplItemId`, `TraitItemId`, etc.

File: compiler/rustc_middle/src/hir/map/mod.rs (hunk 11 above)
Comment (kckeiks) [reply compiler/rustc_middle/src/hir/map/mod.rs:166]: I'm going to rename this method.

And do you mean this?
```rust
    pub fn for_each_module(self, f: impl Fn(LocalDefId)) {
        let crate_items = self.tcx.hir_crate_items(());
        for module in crate_items.submodules.iter() {
            f(*module)
        }
    }

    #[cfg(not(parallel_compiler))]
    #[inline]
    pub fn par_for_each_module(self, f: impl Fn(LocalDefId)) {
        self.for_each_module(f)
    }
```
If so, we would need a `for_each_item` method. We might end up with something like this?

```rust
    pub fn for_each_item(self, f: impl Fn(ItemId)) {
....
    }

    #[cfg(not(parallel_compiler))]
    #[inline]
    pub fn par_for_each_item(self, f: impl Fn(ItemId)) {
        self.for_each_item(f)
    }

    #[cfg(parallel_compiler)]
    #[inline]
    pub fn par_for_each_item(self, f: impl Fn(ItemId)) {
        par_for_each_in(...)
    }
```

but `par_for_each_in` has conditional compilation depending on `parallel_compiler` which would go unused.
Sorry, If I'm misunderstanding something. And maybe we can work on this parallelization on a follow up PR?

File: compiler/rustc_hir/src/target.rs (hunk 5 above)
Comment (kckeiks) [reply compiler/rustc_hir/src/target.rs:153]: I added a comment so that users are aware of this until we merge `Target` and `DefKind`.

**Comment by cjgillot** [comment]
This PR is the first in a series that should significantly reduce full-HIR traversal by queries. This is expected to bring improvements to incremental compilation, by decreasing result invalidation.  This first PR creates the infrastructure for this change, as a new query `hir_crate_items`.  This query does traverse the full HIR to record all definitions in order of appearance, so is quite slow to compute.  The overall regression is 0.5% on primary benchmarks, this should be compensated when follow-up PRs land.
@rustbot label: +perf-regression-triaged

@bors r+

**Pull request merged**